staging/src/k8s.io/api/autoscaling/v2beta2/types.go

```diff
@@ -373,6 +373,7 @@ type HorizontalPodAutoscalerStatus struct {
 
 	// conditions is the set of conditions required for this autoscaler to scale its target,
 	// and indicates whether or not those conditions are met.
+	// +optional
 	Conditions []HorizontalPodAutoscalerCondition `json:"conditions" protobuf:"bytes,6,rep,name=conditions"`
```

Comment:
Other optional slices are annotated with `json:omitempty` and `protobuf:opt`.  This one should be too.

We should also ask for API review of this change, in particular for backward compatability.  E.g. what happens when you use serialize an HPA with `nil` conditions?  Before it would produce an empty slice.  Afterward it would produce a `null`.  Does that mean a new version of kubectl might send an invalid protobuf to a previous server version?  I don't think you can safely make a protobuf field optional after the fact.

Reply:
I'm fine with adding `json:omitempty`, but I don't think it should be annotated with `protobuf:opt` here. The current annotation `protobuf:rep` will allow this field to be repeated any number of times including zero, while `protobuf:opt` will allow zero or one of this field but not more, which sounds more like a breaking change. (Referencing https://developers.google.com/protocol-buffers/docs/proto#specifying_field_rules)

In addition, other optional slices here (HorizontalPodAutoscalerSpec.Metrics, HPAScalingRules.Policies) have `json:omitempty` but not `protobuf:opt`. (Note that the above HorizontalPodAutoscalerStatus.CurrentMetrics actually has neither tag, but that isn't the concern here.)

Reply:
You are totally right about the `protobuf:rep` tag.  I overlooked that.

I think all the slices should have `json:omitempty`.  Policies does, just CurrentMetrics is missing.  Will you please add it?  Otherwise this is good-to-go.

Reply:
Yep! Wanted to make sure we didn't need any other changes. Added the `json:omitempty` tag.